hazelcast/src/main/java/com/hazelcast/internal/server/tcp/LocalAddressRegistry.java

```diff
@@ -0,0 +1,271 @@
+     * To make sure not delete a new entry of a rejoined member with the same uuid
+     * by a stale connection close event, both member uuid and primary address of
+     * the connection that was closed is checked if it matches with the entry inside
+     * the address registry.
+     * <p>
+     * If the uuid and addresses match some entry in the address registry, we try to
+     * remove this entry if the all connections belong to this member is closed. We
+     * keep track the number of active connections belongs to this member uuid entry
+     * and remove if there is no active connection to this member uuid.
+     *
+     * @param instanceUuid    instance uuid of which we try to remove its registration entry
+     * @param primaryAddress  primary address which is set as a Connection#remoteAddress
+     *                       to remove the registration entry
+     */
+    public void tryRemoveRegistration(@Nonnull UUID instanceUuid, @Nonnull Address primaryAddress) {
+        uuidToAddresses.computeIfPresent(instanceUuid, (uuid, linkedAddressesRegistrationCountPair) -> {
+            LinkedAddresses addresses = linkedAddressesRegistrationCountPair.getAddresses();
+            if (addresses.contains(primaryAddress)) {
+                AtomicInteger registrationCount = linkedAddressesRegistrationCountPair.getRegistrationCount();
+                // there is no active connection after this remove the entry
+                if (registrationCount.decrementAndGet() == 0) {
+                    // not using removeIf due to https://bugs.java.com/bugdatabase/view_bug.do?bug_id=8078645
+                    Iterator<UUID> iterator = addressToUuid.values().iterator();
+                    while (iterator.hasNext()) {
+                        UUID currUuid = iterator.next();
+                        if (currUuid.equals(instanceUuid)) {
+                            iterator.remove();
+                        }
+                    }
+                    if (logger.isFineEnabled()) {
+                        logger.fine(addresses + " previously registered for the instance uuid=" + instanceUuid
+                                + " are removed from the registry");
+                    }
+                    // remove the entry
+                    return null;
+                }
+            }
+            return linkedAddressesRegistrationCountPair;
+        });
+    }
+
+    /**
+     * If this address or its resolved IP address has been registered before,
+     * it returns the instance uuid corresponding to this address.
+     * @param address address of hz instance
+     * @return the registered instance uuid corresponds to given address,
+     *          null if the address is not registered
+     */
+    @Nullable
+    public UUID uuidOf(@Nonnull Address address) {
+        try {
+            // try to resolve this address first
+            address = new Address(address.getInetSocketAddress());
```

Comment:
This method is in the operation execution hot path, potentially introduces a hostname lookup? Can it be avoided?

Reply:
Avoided the resolution. Where necessary, perform this resolving before passing it into this method.